crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1428,7 +1433,30 @@ impl<'db> Type<'db> {
                         .member(db, "__dict__");
                 }
 
-                let global_lookup = symbol(db, global_scope(db, *file), name);
+                // If the file that originally imported the module has also imported a submodule
+                // named [name], then the result is (usually) that submodule, even if the module
+                // also defines a (non-module) symbol with that name.
+                //
+                // Note that technically, either the submodule or the non-module symbol could take
+                // priority, depending on the ordering of when the submodule is loaded relative to
+                // the parent module's `__init__.py` file being evaluated.  That said, we have
+                // chosen to always have the submodule take priority.  (This matches pyright's
+                // current behavior, and opposite of mypy's current behavior.)
+                if let Some(submodule_name) = ModuleName::new(name) {
+                    let importing_file = module_ref.importing_file(db);
+                    let importing_index = semantic_index(db, importing_file);
+                    let mut full_submodule_name = module_ref.module(db).name().clone();
```

Comment:
Hmm, I'm a bit concerned about calling `semantic_index` here. Are there cases where `importing_file != self.file`? If so, then we shouldn't call `semantic_index` here because it would introduce a cross-file dependency: This file would always be rechecked if the `importing_file`'s AST changes.

Reply:
Never mind. I thought we're in the `SemanticindexBuilder` but that's not the case. So I think this is fine.